exchange/utils.go

```diff
@@ -231,6 +257,49 @@ func (rs *requestSplitter) cleanOpenRTBRequests(ctx context.Context,
 	return
 }
 
+// cloneBidderReq - clones bidder request and replaces req.User and req.Device with new copies
+func cloneBidderReq(req *openrtb2.BidRequest) *openrtb_ext.RequestWrapper {
+
+	// bidder request may be modified differently per bidder based on privacy configs
+	// new request should be created for each bidder request
+	// pointer fields like User and Device should be cloned and set back to the request copy
+	var newReq *openrtb2.BidRequest
+	newReq = ptrutil.Clone(req)
+
+	if req.User != nil {
+		var userCopy *openrtb2.User
+		userCopy = ptrutil.Clone(req.User)
+		newReq.User = userCopy
+
+		if req.User.Geo != nil {
+			var userGeoCopy *openrtb2.Geo
+			userGeoCopy = ptrutil.Clone(req.User.Geo)
+			newReq.User.Geo = userGeoCopy
+		}
+	}
+
+	if req.Device != nil {
```

Comment:
Please add `ortb.CloneDevice` and call it from here.

Reply:
There are no `CloneDevice` and `CloneSource` functions. Is there a reason not to have them? Should I add them?

Reply:
Clone functions added